### DISCUSSION TITLE
Updates to app and env templates

### DISCUSSION THREAD

**Discussion by buddhike** [aws/copilot-cli#2600] [Q&A] [answered]
Imagine this scenario:

- Create an app, env and svc with copilot version 1.0
- Deploy a newer version of svc with copilot version 1.1

If copilot 1.1 has introduced changes to app and env templates that are required for services to operate correctly, would `copilot svc deploy` command alert the user about this?

**Comment by buddhike** [comment, marked as answer]
It looks like deploy command runs env upgrade.
https://github.com/aws/copilot-cli/blob/mainline/internal/pkg/cli/svc_deploy.go#L162

Does it do that for app template as well?

Also, if the upgrade was triggered by two `copilot svc deploy` commands for different services in same app, does it handle the scenario gracefully? 😊

**Reply by efekarakus** [reply]
> Does it do that for app template as well?

Copilot will check if the app template is out-of-date if so it will ask the user to run `copilot app upgrade` first.

> Also, if the upgrade was triggered by two copilot svc deploy commands for different services in same app, does it handle the scenario gracefully? 😊

Yes, in case there are multiple services that are is trying to touch the environment stack. Only one of them will be able to update it while the others will be stuck waiting. In case one of the services timeout due to waiting (~15 mins) an error is thrown and the `deploy` command can be safely retried!

**Reply by buddhike** [reply]
Thanks @efekarakus 🙏🏾 . This makes sense.
